### PR TITLE
Metric spike - what we want

### PR DESCRIPTION
**Pull request by mfcrocker**

### CODE REVIEW

**Comment by ChristopherMShelton** [comment]
Gunna need another screen

**Comment by mfcrocker** [comment]
> Gunna need another screen

Absolutely! That's a job for the OAT 😉

File: doc/reports/spikes/metrics-what-we-want-spike.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# Monitoring Performance - What should we monitor?
+## Introduction
+This spike report follows on from work and discussions in the OAT meeting to identify areas where we want to be monitoring the system. This outlines what I see as an idealised monitoring setup - there may be blockers to overcome to achieve some of this. I've also provided guidance in the document to how I see the monitoring being provided and used.
+## Production metrics
+All of these metrics should be radiated on monitors
+### Performance
+* **99% call time** - This can be pulled from Google Analytics, CloudWatch or Zipkin headers in Splunk logs, potentially further broken down by microservice
```
Comment (insacuri) [inline doc/reports/spikes/metrics-what-we-want-spike.md:14]: Might be a stupid question but what does this bullet point mean?

Comment (mfcrocker) [reply doc/reports/spikes/metrics-what-we-want-spike.md:14]: 99% call time is the time below which 99% of page calls are served up. Our self-imposed non-functional requirement is that this is below 2 seconds :)

File: doc/reports/spikes/metrics-what-we-want-spike.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
+# Monitoring Performance - What should we monitor?
+## Introduction
```
Comment (necrophonic) [inline doc/reports/spikes/metrics-what-we-want-spike.md:4]: 🎨 Markdown linting - should have blank line before and after headings

File: doc/reports/spikes/metrics-what-we-want-spike.md [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+This spike report follows on from work and discussions in the OAT meeting to identify areas where we want to be monitoring the system. This outlines what I see as an idealised monitoring setup - there may be blockers to overcome to achieve some of this. I've also provided guidance in the document to how I see the monitoring being provided and used.
+## Production metrics
+All of these metrics should be radiated on monitors
+### Performance
```
Comment (necrophonic) [inline doc/reports/spikes/metrics-what-we-want-spike.md:12]: Might be worth looking at these metric as SLIs (Service Level Indicators) - would then need to establish what the SLOs (Service Level Objectives) for each actually are (though gathering baselines is the first step)

File: doc/reports/spikes/metrics-what-we-want-spike.md [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+* **99% call time** - This can be pulled from Google Analytics, CloudWatch or Zipkin headers in Splunk logs, potentially further broken down by microservice
+* **Current concurrent users** - This can be pulled most easily from Google Analytics
```
Comment (necrophonic) [inline doc/reports/spikes/metrics-what-we-want-spike.md:15]: Need to define what "concurrent users" means

File: doc/reports/spikes/metrics-what-we-want-spike.md [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+* **Peak concurrent users** - As above, but the maximum for the last 24h/72h/7d
+### Physical
+* **Diego cell usage** - This can be pulled from Cloudwatch. It's a candidate for alerting into Slack (and does currently alert), but note that Diego Cell balancing can often be poor (consider alerting if a significant number of DCs are maxed or the average load exceeds x%)
```
Comment (necrophonic) [inline doc/reports/spikes/metrics-what-we-want-spike.md]: Short term and will be replaced when moving off the current infrastructure. Could investigate using https://github.com/sambryant4/diego-capacity-monitor (running as a demo in devtest currently)

**Comment by necrophonic** [comment]
+1 for extra screen space. Genuinely a good idea as a single rotating screen (of small size) is not sufficient to correctly radiate status for a production system.

File: doc/reports/spikes/metrics-what-we-want-spike.md [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+* **Database CPU%** - This can be pulled from Cloudwatch or by passing DB physical metrics into Splunk. This is a very good candidate for alerting into Slack, as going above 80% is usually a precursor to a collapse in system performance
+### Platform 
+* **App crashes** - This can be pulled from CloudFoundry or Splunk. It can be a candidate for alerting, but would need to be set to an appropriate threshold - single app crashes are usually fine and don't require investigation while multiple in short order need flagging up
```
Comment (rhiGriff) [inline doc/reports/spikes/metrics-what-we-want-spike.md]: Does this include app downtime? If so I think we need to be clearer here. We will also want to monitor if the internal site (ROps) is not available too as we currently alert when frontstage is down.

**Review by rhiGriff — COMMENTED** [review]
We also need to consider monitoring the Rabbit DLQs too.

File: doc/reports/spikes/metrics-what-we-want-spike.md (hunk 1 above)
Comment (owencjones) [reply doc/reports/spikes/metrics-what-we-want-spike.md:14]: So it's a 99% success rate?

Actually seems rather low, tbh, but it's a start!

File: doc/reports/spikes/metrics-what-we-want-spike.md (hunk 2 above)
Comment (owencjones) [reply doc/reports/spikes/metrics-what-we-want-spike.md:4]: JG knows his MD

File: doc/reports/spikes/metrics-what-we-want-spike.md (hunk 6 above)
Comment (owencjones) [reply doc/reports/spikes/metrics-what-we-want-spike.md]: Further to Rhi's point, we need to check that healthcheck endpoints are are reflective of health, as I think recently one was reporting healthy when it wasn't.

File: doc/reports/spikes/metrics-what-we-want-spike.md (hunk 1 above)
Comment (owencjones) [reply doc/reports/spikes/metrics-what-we-want-spike.md:14]: Are we due NFRs from anyone?  Blue sky NFRs are often not useful.

**Review by owencjones — COMMENTED** [review]
Only not approved because I've made comments to consider.  Don't consider this a blocker - although others' comments obviously may need addressing.

**Review by owencjones — DISMISSED** [review]
Approved on the caveat that you will address others' comments.

File: doc/reports/spikes/metrics-what-we-want-spike.md (hunk 1 above)
Comment (mfcrocker) [reply doc/reports/spikes/metrics-what-we-want-spike.md:14]: Na, not to be confused with 99(.9999999)% uptime. This is simply the time that 99% of users can expect to get, as opposed to the median (or 50% call time)

File: doc/reports/spikes/metrics-what-we-want-spike.md (hunk 3 above)
Comment (mfcrocker) [reply doc/reports/spikes/metrics-what-we-want-spike.md:12]: I support this, but it's beyond the scope of this spike (and is a job for OAT)

File: doc/reports/spikes/metrics-what-we-want-spike.md (hunk 4 above)
Comment (mfcrocker) [reply doc/reports/spikes/metrics-what-we-want-spike.md:15]: Presumably concurrent users would be defined by whichever platform we're pulling this metric from, so it's hard for me to provide a specific definition here.

File: doc/reports/spikes/metrics-what-we-want-spike.md (hunk 5 above)
Comment (mfcrocker) [reply doc/reports/spikes/metrics-what-we-want-spike.md]: A good point :)

File: doc/reports/spikes/metrics-what-we-want-spike.md (hunk 6 above)
Comment (mfcrocker) [reply doc/reports/spikes/metrics-what-we-want-spike.md]: We could measure app downtime if we felt it useful but it's not what I'm raising here. I'm suggesting that any app that crashes more than a couple of times in an hour should be alerted and investigated.

**Review by owencjones — APPROVED** [review]

**Pull request merged**